### PR TITLE
Bloom upgrade in master

### PR DESCRIPTION
**Pull request by NikVolf**
had to move the same in-place upgrade logic that we had in beta

### CODE REVIEW

**Comment by rphmeier** [comment]
should add a comment warning about the dangers of adding future migrations which affect the bloom without adjusting the inplace migrations.

File: ethcore/src/migrations/upgrade_bloom.rs [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+					if idx % 10000 == 1 { progress.tick(); };
+				}
+
+				LittleEndian::write_u64(&mut key, ACCOUNT_BLOOM_SPACE as u64);
+				batch.put(COL_ACCOUNT_BLOOM, ACCOUNT_BLOOM_SPACE_KEY, &key);
+			},
+		}
+	}
+
+	println!("Adding/expanding accounts bloom (one-time upgrade)");
```
Comment (rphmeier) [inline ethcore/src/migrations/upgrade_bloom.rs]: same `println!` issue as the beta PR.

File: ethcore/src/migrations/upgrade_bloom.rs [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+	let bloom_journal = {
+		let mut bloom = Bloom::new(ACCOUNT_BLOOM_SPACE, DEFAULT_ACCOUNT_PRESET);
+		// no difference what algorithm is passed, since there will be no writes
+		let state_db = journaldb::new(
```
Comment (rphmeier) [inline ethcore/src/migrations/upgrade_bloom.rs]: this comment is actually wrong -- `OverlayRecent` will work with `Archive`, but not vice-versa, as `Archive` wouldn't be able to read the latest state from the journal overlay.

Comment (rphmeier) [inline ethcore/src/migrations/upgrade_bloom.rs]: It's also probably best to just take the pruning argument as a parameter -- we have it in the migration phase, so there's no point firing shots in the dark. I don't see why `refcounted` and `light` pruning wouldn't be supported with this migration.

**Comment by NikVolf** [comment]
@rphmeier i don't see where exactly this warning will have any use

**Comment by rphmeier** [comment]
@NikVolf If you imagine a migration `10 -> 11` which assumes the bloom is adjusted to a larger size, it'll fail for every `9 -> 11` migration if the inplace migration isn't adjusted accordingly. The warning will help the developer write the code correctly and also signal that the inplace migrations as a temporary mechanism have outstayed their welcome.

File: ethcore/src/migrations/upgrade_bloom.rs (hunk 2 above)
Comment (NikVolf) [inline ethcore/src/migrations/upgrade_bloom.rs]: why it will fail with this error?
there is only one state used there, which is for the best block, no?

**Comment by NikVolf** [comment]
@rphmeier 
no, if we alter bloom size sometimes, such in-place upgrades are not temporary mechanism
i can't see how 9->11 migration will fail here
bloom will just be generated with the most recent options after 9->11 migration will finish

**Comment by rphmeier** [comment]
Ok, maybe you can explain the flaw in the reasoning here, but the way I see it is:

1. 9 -> 10 migration applied, creating a bloom with the initial size.
2. 10 -> 11 migration applied, which expects a bloom with the new size and fails, because inplace migrations are applied at the end.

**Comment by NikVolf** [comment]
@rphmeier 
9 -> 10 migration will create bloom with the most actual size right away, so this is not the case

**Comment by NikVolf** [comment]
Looks like trie iterations take too long on many machines for 10M+ accounts

**Comment by gavofyork** [comment]
closing for now; probably worth keeping the code in general since we'll want to rebuild after we've done the pruning and after a sufficiently high number of suicides have happened.

**Pull request closed without merging**